### ISSUE TITLE
how to turn high-dimension data like (256, 256, 4)  into .rec file

### ISSUE THREAD

**Issue by monaen** [apache/incubator-mxnet#2987]
I have some data and each item is in (256, 256 ,4) format, and I want to know how to get the .rec files based on these data. Any ideas? thanks !!!

**Comment by piiswrong** [comment]
mx.recordio.MXRecordIO

**Comment by monaen** [comment]
Hi, piiswrong
Would you please provide a more detailed example about how to use mx.recordio.MXRecordIO?

**Comment by Ldpe2G** [comment]
@monaen https://github.com/dmlc/mxnet/blob/master/tools/im2rec.py

**Comment by monaen** [comment]
Thanks you all,
@Ldpe2G :   I follow the example you provied, and write a program to generate rec database:

from collections import namedtuple
import struct
IRHeader = namedtuple('HEADER', ['flag', 'label', 'id', 'id2'])
_IRFormat = 'IfQQ'
_IRSize = struct.calcsize(_IRFormat)


prefix='train'
record = mx.recordio.MXRecordIO(prefix+'.rec', 'w')

Head_item = zip(np.zeros(len(y_train)).astype(np.int), y_train, range(len(y_train)), np.zeros(len(y_train)).astype(np.int))
print Head_item[0]
### HEADER(flag=0, label=2, id=0, id2=0)


### my original train dataset is:   X_train.shape = (1900,4,256,256)
### and I vectoring each item to a vector using ravel()

for i in range(len(X_train)):
    _flag, _label, _id, _id2 = Head_item[i]
    header = mx.recordio.IRHeader(_flag, _label, _id, _id2)
    data = X_train[i,:]                                             ###############   X_train.shape = (1900, 65536)
    s = pack_data(header, data)
    record.write(s)


### I generate the val and test dataset the same way, and 
### Then I put the rec data into my model, however I got an error
### This is my train_model.py

def get_symbol(num_classes = 5):
    input_data = mx.symbol.Variable(name="data")
   ......... .........
   ......... .........
    return cellnet

cellnet = get_symbol()


batch_size = 38
total_batch = 1900 / 38 

data_shape = (4, 305, 305)
train_iter = mx.io.ImageRecordIter(
	shuffle=True,
    path_imgrec="Datasets/train.rec",
    mean_img="Datasets/mean.bin",
    data_shape  = data_shape,
    batch_size  = batch_size,
    rand_crop   = True,
    rand_mirror = True,
    preprocess_threads=1)

test_iter = mx.io.ImageRecordIter(
    path_imgrec="Datasets/val.rec",
    mean_img="Datasets/mean.bin",
    rand_crop   = False,
    rand_mirror = False,
    data_shape  = data_shape,
    batch_size  = batch_size,
    preprocess_threads=1)

num_epoch = 100

model = mx.model.FeedForward(
    ctx = mx.gpu(0),      # Run on GPU 0
    symbol = cellnet,         # Use the network we just defined
    num_epoch = 10,       # Train for 10 epochs
    learning_rate = 0.1,  # Learning rate
    momentum = 0.9,       # Momentum for SGD with momentum
    wd = 0.00001)         # Weight decay for regularization
model.fit(
    X=train_iter,  # Training data set
    eval_data=test_iter,  # Testing data set. MXNet computes scores on test set every epoch
    batch_end_callback = mx.callback.Speedometer(batch_size, 200))  # Logging module to print out progress


### the error is:
/home/utranus/mxnet/dmlc-core/include/dmlc/logging.h:235: [13:01:38] src/io/image_aug_default.cc:254: Check failed: static_cast<index_t>(res.rows) >= param_.data_shape[1] && static_cast<index_t>(res.cols) >= param_.data_shape[2] input image size smaller than input shape
terminate called after throwing an instance of 'dmlc::Error'
  what():  [13:01:38] src/io/image_aug_default.cc:254: Check failed: static_cast<index_t>(res.rows) >= param_.data_shape[1] && static_cast<index_t>(res.cols) >= param_.data_shape[2] input image size smaller than input shape
Aborted (core dumped)


The data shape estimated to be the question, but I do not know where specifically is going wrong...

Thanks

**Comment by piiswrong** [comment]
this pr https://github.com/dmlc/mxnet/pull/3005 enables multi dimensional labels. Use --pack_label=1 to pack label together with data

**Comment by monaen** [comment]
@piiswrong
I check the PR #3005 but I still cannot figure out.

Do you mean I can generate rec database using im2rec.py ? 
like if I want to generate train.rec file, I should first save my X_train ( with dimension(1900,4,305,305) )  to 1900*4 images and use im2rec.py with --pack_label=1 to get my train.rec?

**Comment by monaen** [comment]
@piiswrong
Actually, I can now run the model and load data using **mx.io.NDArrayIter**, but it seems a little bit slowly than using rec database I think

**Comment by machuck** [comment]
What is the latest status of high dimensional data? I am looking for an arbitrary number of dimensions, because the problem is not image data. How does MXNET work with 4,5,6... dimensional data in general? Thanks!

**Comment by piiswrong** [comment]
http://mxnet.io/tutorials/python/record_io.html

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
